Notebook to compare ESTIMATE to consensus cell types

sjspielman opened the pull request:
Closes ~#154~ #159 
Stacked on #158 

This PR adds a notebook to compare ESTIMATE scores, estimated with either pseudobulk or bulk counts, to the consensus cell types. I first classify cells into "immune", "stromal", or "other" categories (note I have a few TODOs here about some of those cell type choices!). Then I look at the proportions of these categories across samples, and compare those proportions to estimate scores mostly with scatterplots. I've added some test directly interpreting the figures, but still thinking about what some of these results mean (or how meaningful they are) so I don't have any "overall" conclusions here.

I also added a script to cp (not sync 😑 ) consensus cell type files from the staging bucket for now. 


Here's a copy of the notebook! Let me know what additional figures you think would be helpful to include here (and thereby include also for group discussion next sprint). [compare-estimate-consensus.nb.html.zip](https://github.com/user-attachments/files/18893545/compare-estimate-consensus.nb.html.zip)

jashapiro commented:
> I also added a script to cp (not sync 😑 )

Why the face? What problem did you run into with sync?

sjspielman commented:
> Why the face? What problem did you run into with sync?

When I ran the shell script with `aws s3 sync` instead of copy, the consensus cell type files downloaded didn't have neurons and were an old version of those files. Such a thing made no sense to me... I must have done something weird without realizing it, so I saved myself from myself by using `cp` instead.

jashapiro commented:
> > Why the face? What problem did you run into with sync?
> 
> When I ran the shell script with `aws s3 sync` instead of copy, the consensus cell type files downloaded didn't have neurons and were an old version of those files. Such a thing made no sense to me... I must have done something weird without realizing it, so I saved myself from myself by using `cp` instead.

It looks like [5e2caa4](https://github.com/AlexsLemonade/scpca-paper-figures/pull/160/commits/5e2caa420bb13955f7890198c6544a207a4e1b2a) also changed the source bucket, was that not the issue?

sjspielman commented:
> It looks like [5e2caa4](https://github.com/AlexsLemonade/scpca-paper-figures/pull/160/commits/5e2caa420bb13955f7890198c6544a207a4e1b2a) also changed the source bucket, was that not the issue?

It was definitely at first! Then I deleted all the files, synced from staging, and it happened again. Honestly I think the most likely thing that happened is probably didn't save the file with enough ooomph and passion and purpose hitting those ctl+s keys, and the code I thought I was running isn't what I was running. But, I don't have a sense that this script will stick around either way indefinitely, since we will want to get consensus cell types for figures more generally, not only this analysis. This script just allows us to start this analysis here for now as I see it.

allyhawkins commented on analysis/estimate/notebooks/compare-estimate-consensus.Rmd:
````diff
@@ -0,0 +1,354 @@
+```
+There appears to be a negative relationship for SCPCP000001 and SCPCP000017, a very weak negative relationship for SCPCP000006 (possibly outlier driven?), and no relationship for SCPCP000002 or SCPCP000009.
+
+## Relationship between cell types and bulk ESTIMATE scores
+
+We'll do the same comparisons here but with ESTIMATE scores estimated from bulk counts (not bulk TPM) instead of from pseudobulk.
````
I don't necessarily expect these to be correlated. If our hypothesis is that we get different cell compositions between bulk and single-cell, then it's likely that these plots will look different than the ones comparing ESTIMATE to pseudobulk.

allyhawkins commented:
> I'm not sure I follow what you are envisioning in terms of a scatterplot here. In terms of correlation, there is only 1 value per project, so there are 5 values per estimate score type - and thereby at most 5 points in a plot that shows a summary statistic. Are you suggesting to look at the relationship between stromal and immune correlations?

I would make a plot with one point per sample with pseudobulk ESTIMATE scores on one axes and fraction of cells on the other axes. The same plots that you already have but without faceting by project and then add a correlation coefficient. You should make one plot for immune, one for stromal, and one for other cells with the caveat that I really only expect to see strong correlation in the immune one.

sjspielman commented:
Plot colors should be more visible now, and I added a section with scatterplots as requested. Solid hypothesis 👍 
[compare-estimate-consensus.nb.html.zip](https://github.com/user-attachments/files/18914453/compare-estimate-consensus.nb.html.zip)

allyhawkins commented on analysis/estimate/notebooks/compare-estimate-consensus.Rmd:
````diff
@@ -0,0 +1,340 @@
+For human readability, we'll look at the annotations, not ontologies, for this.
+
+```{r}
+all_celltypes <- celltype_data |>
````
The rendered copy has the list printing out twice, but I'm not sure why given the code. Also can you sort these alphabetically?

sjspielman replied:
> The rendered copy has the list printing out twice, but I'm not sure why given the code.

This is a long-standing very irritating bug with rendering rmarkdown notebooks. Like, it's all over our rendered training notebooks too 😭 😭 😭 😭

sjspielman replied:
But yes, will sort!!

allyhawkins replied:
> > The rendered copy has the list printing out twice, but I'm not sure why given the code.
> 
> This is a long-standing very irritating bug with rendering rmarkdown notebooks. Like, it's all over our rendered training notebooks too 😭 😭 😭 😭

Ah that makes sense! I do remember that being an issue.

allyhawkins commented on analysis/estimate/notebooks/compare-estimate-consensus.Rmd:
```diff
@@ -0,0 +1,340 @@
+
+* Nearly all samples across SCPCP000001, SCPCP000002, and SCPCP000006 lack "stromal cells" 
+  * From the plot, it appears only 1 sample in SCPCP000001 and 1 sample in SCPCP000006 have stromal; if other samples have, it's an exceedingly small proportion
+* Only SCPCP000006 and SCPCP000009 have samples with truly visible levels of "stromal cells", particularly SCPCP000009
```
```suggestion
* Only SCPCP000006 and SCPCP000017 have samples with truly visible levels of "stromal cells", particularly SCPCP000009
```
I think this should be 17?

allyhawkins commented on analysis/estimate/notebooks/compare-estimate-consensus.Rmd:
````diff
@@ -0,0 +1,340 @@
+  theme(axis.text.x = element_blank()) 
+```
+
+
````
Let's add some comments about the immune cells. To be honest, I don't really trust the stromal scores or stromal cell type distributions as much as I do the immune cells, so I would focus on that.

allyhawkins commented on analysis/estimate/notebooks/compare-estimate-consensus.Rmd:
```diff
@@ -0,0 +1,340 @@
+
+In theory, the `Unknown` cells in our consensus annotation are more likely to be tumor cells.
+We also excluded several cell types from analysis which might also be tumor.
+Similarly in theory, the ESTIMATE score (sum of immune and stromal) is expected to be proportional to tumor purity.
```
```suggestion
Similarly in theory, the ESTIMATE score (sum of immune and stromal) is expected to be proportional to tumor purity where higher ESTIMATE scores correspond to lower tumor purity. 
Therefore we expect there to be a negative correlation between fractions of cells that are non-immune/non-stromal and the ESTIMATE score. 
```
Let's just be explicit about what this metric is looking at since I think this concept can be confusing.

sjspielman replied:
Oooh indeed thank you, as I was confusing myself too! Negative indeed does make sense. phew.

allyhawkins commented on analysis/estimate/notebooks/compare-estimate-consensus.Rmd:
```diff
@@ -0,0 +1,340 @@
+
+There appears to be a negative relationship for SCPCP000001 and SCPCP000017, a very weak negative relationship for SCPCP000006 (possibly outlier driven?), and no relationship for SCPCP000002 or SCPCP000009.
+
+This may reflect that the genes in ESTIMATE do not comprehensively capture all non-cancer genes, and/or include genes that are highly expressed in some of our tumor cells.
```
This also could be a limitation of consensus cell types as stromal cells are more similar to cancer cells and hard to distinguish when cell typing.

allyhawkins commented on analysis/estimate/notebooks/compare-estimate-consensus.Rmd:
```diff
@@ -0,0 +1,340 @@
+We'll next visualize the same results, but pooling all projects together and with correlation coefficients (both $R$ and $\rho$).
+We'll also color points based on whether the pseudobulk was derived from single-cell or single-nuclei experiment.
+
+Again, note that score shown in the the third plot with the "other" celltypes is the ESTIMATE score, which is the sum of stromal and immune.
```
```suggestion
Again, note that the score shown in the the third plot with the "other" celltypes is the ESTIMATE score, which is the sum of stromal and immune.
```

allyhawkins commented on analysis/estimate/notebooks/compare-estimate-consensus.Rmd:
````diff
@@ -0,0 +1,340 @@
+)
+```
+
+We see the strongest correlation for immune celltypes as expected, this plot also emphasizes the trend that single-cell show higher fractions of immune cells than do single-nuclei samples.
````
I think I might also note that we see the correct direction for all three plots where immune and stromal show positive relationships and other shows a negative relationship.

allyhawkins commented on analysis/estimate/sync-consensus-celltypes.sh:
```diff
@@ -0,0 +1,24 @@
+
+for project in SCPCP000001 SCPCP000002 SCPCP000006 SCPCP000009 SCPCP000017; do
+  mkdir -p ${consensus_dir}/${project}
+  aws s3 cp s3://openscpca-nf-workflow-results-staging/2024-11-25/cell-type-consensus/${project}/ ${consensus_dir}/${project} --profile ${aws_profile} --recursive
```
```suggestion
  aws s3 cp s3://openscpca-nf-workflow-results/2024-11-25/cell-type-consensus/${project}/ ${consensus_dir}/${project} --profile ${aws_profile} --recursive
```
Just noting that the updated results are now in the actual results bucket and not on staging, so you could change this.

allyhawkins reviewed (approved):
A few smaller comments, mostly about interpretation that should be addressed, but I don't need to see this again.

sjspielman commented:
@allyhawkins while implementing reviews for this, I had a thought I think is important! We had previously identified a set of low-quality samples to remove: https://github.com/AlexsLemonade/scpca-paper-figures/blob/90563f229910042f501576752992296de8925986/analysis/pseudobulk-bulk-prediction/notebooks/build-assess-models.Rmd#L41-L53

These didn't make it into the last notebook, but perhaps it should make it into this notebook and we should exclude these samples from scatterplots (and correlations!) comparing between estimate and cell type fractions. Alternatively, we can make 2 versions of plots. I'd also want to point out in the cell fraction barplots which samples are low-quality. This could (should?) also be a subsequent PR too. What do you think all around here?

allyhawkins commented:
> These didn't make it into the last notebook, but perhaps it should make it into this notebook and we should exclude these samples from scatterplots (and correlations!) comparing between estimate and cell type fractions. Alternatively, we can make 2 versions of plots. I'd also want to point out in the cell fraction barplots which samples are low-quality. This could (should?) also be a subsequent PR too. What do you think all around here?

Yes we should exclude these in the scatterplots. But I think that might be a subsequent PR. I would merge this in and then update it with the points you mentioned above.
